americangut/diversity_analysis.py

```diff
@@ -0,0 +1,1729 @@
+    show_p_value : bool, optional
+        When True, the signigance bars will display the actual p value.
+        Otherwise, p values will be coded as (p <= 0.1: '+', p <= 0.05: '*',
+        p <= 0.01: '**', p <= 0.001: '***', p <= 0.0001).
+    show_xgrid: bool, optional
+        Default is False. Adds vertical lines at each major x-tick.
+    show_ygrid: bool, optional
+        Default is True. Adds horizonal lines at each major y-tick.
+    title: str, optional
+        The title to be placed on the graph.
+    ylims : list
+        The limits for the y-axis.
+    ylabel : str
+        The label text for the y-axis. Every time you leave off appropriate
+        labels and units, a science grad student grading lab reports cries
+        another bitter tear into their bottle of craft beer.
+
+
+    Returns
+    -------
+    ax : axes
+        A matplotlib axes containing the plotted data
+    """
+
+    # Removes any undefined groups
+    map_ = meta.groupby(meta[group].apply(_check_strs)
+                        ).get_group(True)
+    dm = dm.filter(map_.index)
+
+    # Orders the objects
+    if order is None:
+        order = map_.groupby(group).groups.keys().sort()
+
+    # Gets the reference groups
+    if isinstance(ref_groups, str):
+        ref_groups = [ref_groups]
+    if ref_groups is None:
+        ref_groups = order
+
+    # Checks for an overall signifigant differences
+    if show_p:
+        perma_res = skbio.stats.distance.permanova(dm, map_, group, num_iter)
+        kwargs['p_value'] = perma_res['p-value']
+
+    # Gets the distance vectors
+    w_groups, w_dist, b_groups, b_dist = \
+        get_distance_vectors(dm, meta, group, order)
+
+    bar_counter = 0
+
+    # Loops through the data to make boxplots compared to the reference group
+    for id1, ref_group in enumerate(ref_groups):
+        # Determines the offset
+        offset = (id1 * len(order) + bar_counter)*width
+        # Determines the position of the reference group
+        ref_loc = order.index(ref_group)
+
+        # Adds the distance vector to the means
+        dist_bar = np.zeros((len(order)))
+        dist_std = np.zeros((len(order)))
+
+        dist_bar[ref_loc] = w_dist[w_groups.index(ref_group)].mean()
+        dist_std[ref_loc] = w_dist[w_groups.index(ref_group)].std()
+
+        # Sets up the pvalues
+        if p_table is None:
+            p_values = p_table
+        else:
+            p_values = np.array([])
+            sub_p = pd.concat([p_table.loc[p_table['Group 1'] == '%s vs. %s'
+                                           % (ref_group, ref_group)],
+                               p_table.loc[p_table['Group 2'] == '%s vs. %s'
+                                           % (ref_group, ref_group)]])
+
+        for id2, group in enumerate(order):
+            if group == ref_group:
+                continue
+            # Gets the distance vector
+            dist_bar[id2] = b_dist[b_groups.index({ref_group, group})].mean()
```

Comment:
those `.index` calls are linear searches. It may be worth investigating a more efficient means of retrieval